tests/util/test_check_dependencies.py

```diff
@@ -59,13 +60,28 @@ def test_mandatory_dependency(self) -> None:
             "synapse.util.check_dependencies.metadata.requires",
             return_value=["dummypkg >= 1"],
         ):
-            with self.mock_installed_package(None):
+            with self.mock_installed_package(distribution_with_no_version):
                 self.assertRaises(DependencyException, check_requirements)
+
+    def test_version_reported_as_None(self) -> None:
+        """Complain if importlib.metadata.version() returns None.
+
+        This shouldn't normally happen, but it was seen in the wild (#12223).
+        """
+        with patch(
+            "synapse.util.check_dependencies.metadata.requires",
+            return_value=["dummypkg >= 1; extra == 'mypy'"],
+        ), patch("synapse.util.check_dependencies.RUNTIME_EXTRAS", {"cool-extra"}):
+            # We're testing that none of these calls raise.
+            with self.mock_installed_package(None):
+                check_requirements()
+                check_requirements("cool-extra")
             with self.mock_installed_package(old):
-                self.assertRaises(DependencyException, check_requirements)
+                check_requirements()
+                check_requirements("cool-extra")
             with self.mock_installed_package(new):
-                # should not raise
                 check_requirements()
+                check_requirements("cool-extra")
```

Comment:
I'm very confused by these test changes.

Why did we change `test_mandatory_dependency`? And shouldn't `test_version_reported_as_None` be the test that uses `distribution_with_no_version`?

Reply:
Excellent questions. I'm also very confused.

I think I've accidentally transposed the bodies of `test_version_reported_as_None` and `test_mandatory_dependency`. I was probably rushing :(

Reply:
Apologies again. [83ef7a9](https://github.com/matrix-org/synapse/pull/12244/commits/83ef7a9101cb4f4e30f05c8f2b85d8e4574d2ccf) should hopefully restore this to something sensible.